Add GsaFeedFileSender.

aptls opened the pull request.

Commit:
Add GsaFeedFileSender.java from adaptor library.

Commit:
Modify GsaFeedFileSender for local use.

Update the package. Copy GsaCommunicationHandler.handleGsaException
from the adaptor library into GsaFeedFileSender to handle the use of
this method. Copy IOHelper from the adaptor library and add the guava
jar for the VisibleForTesting annotation.

Commit:
Add GsaFeedFileSender example.

Add an example of using GsaFeedFileSender. Make elements of
GsaFeedFileSender public for use from other packages. Change
methods specific to metadata-and-url feeds to support all feed
types; only a feedtype parameter was missing.

wiarlawd commented on build.xml:
```diff
@@ -5,11 +5,17 @@
   <property name="build.dir" location="build"/>
   <property name="build-src.dir" location="${build.dir}/src"/>
   <property name="generate.dir" location="${build.dir}/generate-src"/>
+  <property name="lib.dir" location="lib"/>
```
Guava counts as third-party, so let's follow the third-party rules (third_party/guava directory with an Apache LICENSE file and the Guava JAR file).

wiarlawd commented on src/com/google/enterprise/gsafeed/examples/SendFeed.java:
```diff
@@ -0,0 +1,52 @@
+  public static void main(String... args) throws Exception {
+    Charset utf8 = Charset.forName("UTF-8");
+
+    String feedFile = args[0];
```
I know this is just a skeleton, but it's easy enough to check args.length and print a usage line.

wiarlawd commented on build.xml:
```diff
@@ -57,4 +64,16 @@
+
+  <target name="send_feed" depends="build"
+      description="Send feed example">
+    <property name="example.feed" value="/full.xml"/>
```
feed becomes feedFile, and gsa becomes gsaHost, and both expanded names are a little more explicable. I'd prefer example.feedfile and example.gsahost.

For Topaz, we included example.args = "${example.feed} ${example.gsa}" to make it easier to override some or all of the command line. Just a suggestion.

wiarlawd commented on build.xml:
```diff
@@ -57,4 +64,16 @@
     </java>
   </target>
+
+  <target name="send_feed" depends="build"
```
Do we want a more gsafeed/records-based name here, to make room to add groups later?

wiarlawd commented on src/com/google/enterprise/gsafeed/examples/SendFeed.java:
```diff
@@ -0,0 +1,52 @@
+public class SendFeed {
+
+  public static void main(String... args) throws Exception {
+    Charset utf8 = Charset.forName("UTF-8");
```
private static final Charset UTF_8?

wiarlawd commented on src/com/google/enterprise/gsafeed/examples/SendFeed.java:
```diff
@@ -0,0 +1,52 @@
+        + " to GSA at " + gsaHost);
+
+    // Read the feed file to get the datasource and feedtype.
+    URL feedUrl = SendFeed.class.getResource(args[0]);
```
s/args[0]/feedFile/

wiarlawd commented on src/com/google/enterprise/gsafeed/examples/SendFeed.java:
```diff
@@ -0,0 +1,52 @@
+    String feedtype = feed.getHeader().getFeedtype();
+
+    // Send the feed to the GSA.
+    GsaFeedFileSender sender = new GsaFeedFileSender(gsaHost, false, utf8);
```
We sometimes add comments to boolean parameters, like /* secure */ false, or add a local variable for self-documentation, boolean secure = false. Same for compression in the line below.

wiarlawd commented on src/com/google/enterprise/gsafeed/IOHelper.java:
```diff
@@ -0,0 +1,154 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.enterprise.gsafeed;
```
This file is unchanged except for the package and the provenance comment?

aptls replied:
Yes.

wiarlawd commented on src/com/google/enterprise/gsafeed/GsaFeedFileSender.java:
```diff
@@ -191,17 +196,16 @@ private void handleGsaReply(String reply, int responseCode) {
    * Datasource name is limited to [a-zA-Z_][a-zA-Z0-9_-]*.
    */
-  void sendMetadataAndUrl(String datasource, String xmlString,
+  public void sendGsaFeed(String datasource, String feedtype, String xmlString,
```
Are you planning to change this later to take a Gsafeed instead?

PapaJoltOmega reviewed (commented):
Please add description to CL.
How much of GsaFeedFileSender.java has changed from library?

Thank you

wiarlawd commented:
There are multiple commits here, and the first commit has GsaFeedFileSender.java unchanged, so you can just look at the second (or second and third) to see the diffs.

PapaJoltOmega commented on src/com/google/enterprise/gsafeed/GsaFeedFileSender.java:
```diff
@@ -34,8 +34,8 @@
-/* Modified from GsaFeedFileSender in adaptor library. */
-class GsaFeedFileSender {
+// Modified from GsaFeedFileSender in adaptor library.
+public class GsaFeedFileSender {
```
why do methods in this class have to become public at this time?

aptls replied:
I wanted to use them in an example. It's possible as we get further along that we'll end up hiding this again in some different interface, but for now I wanted to have something running to verify that this was working outside the adaptor library.

PapaJoltOmega reviewed (approved):
Please merge once John approves. Thanks

Commit:
Code review changes.

Move guava to third_party and add license file.

Use "send_gsafeed" instead of "send_feed" to distinguish this
from a groups example.

aptls committed:
Merge branch 'master' into aptls-sender
